File: src/auth/provider.ts
// This server is its own OAuth authorisation server.
//
// The SDK ships the HTTP half — mcpAuthRouter gives us /register, /authorize,
// /token, /revoke and both metadata documents, with PKCE checking, RFC-shaped
// errors and rate limiting. What it cannot ship is the policy, which is this
// file: who may consent, what a code means, and how long a token lasts.
//
// TOKENS ARE OPAQUE RANDOM STRINGS, NOT JWTs. A JWT would need a signing key,
// which is one more secret in the process that already holds the household's
// credentials, and it would make revocation a list of exceptions rather than a
// delete. Nothing here is stateless enough to be worth that: there is one
// household and one process. See AuthStore for how they are kept.
//
// The one thing to know before changing `authorize()`: its signature is
// (client, params, res) — no `req`. It cannot read a header, so it cannot do
// the exe.dev check itself. That check is middleware, in src/mcp/server.ts,
// which puts the identity on `res.locals` before this ever runs. If you find
// yourself wanting the request here, add to the middleware instead.

import type { Response } from 'express';

import type {
  AuthorizationParams,
  OAuthServerProvider,
} from '@modelcontextprotocol/sdk/server/auth/provider.js';
import type { OAuthRegisteredClientsStore } from '@modelcontextprotocol/sdk/server/auth/clients.js';
import type { AuthInfo } from '@modelcontextprotocol/sdk/server/auth/types.js';
import type {
  OAuthClientInformationFull,
  OAuthTokenRevocationRequest,
  OAuthTokens,
} from '@modelcontextprotocol/sdk/shared/auth.js';
import {
  InvalidGrantError,
  InvalidTokenError,
  ServerError,
} from '@modelcontextprotocol/sdk/server/auth/errors.js';

import { ConsentDesk, consentPage } from './consent.ts';
import { sameEmail, type Identity } from './exedev.ts';
import { AuthStore, newSecret } from './store.ts';

/** An authorisation code is spent in seconds, not minutes. RFC 6749 §4.1.2 says ten. */
const CODE_TTL_SECONDS = 60;

/** An hour. Short enough to matter, long enough that refresh is rare. */
export const ACCESS_TOKEN_TTL_SECONDS = 60 * 60;

export type ProviderOptions = {
  store: AuthStore;
  /** The only email that may approve anything. One household, one entry. */
  owner: string;
  /** Shown on the consent page, so a person can see which folder they are opening. */
  folder: string;
  /**
   * Whether to offer "also connect my Kroger account" on the consent page.
   *
   * `configured` is false when the server has no Kroger credentials, and then
   * the box is not shown at all — a box that cannot do anything is worse than
   * no box. See src/kroger/link.ts for why the link goes here rather than after
   * the authorisation code.
   */
  kroger?: { configured: boolean; connected: () => boolean };
};

export class MealPlanOAuthProvider implements OAuthServerProvider {
  readonly store: AuthStore;
  readonly owner: string;
  readonly desk = new ConsentDesk();
  readonly #folder: string;
  readonly #kroger: { configured: boolean; connected: () => boolean };

  constructor(options: ProviderOptions) {
    this.store = options.store;
    this.owner = options.owner;
    this.#folder = options.folder;
    this.#kroger = options.kroger ?? { configured: false, connected: () => false };
  }

  /**
   * Registration is open, and that is not an oversight — it is what lets an
   * assistant add this server by URL with nobody copying a secret by hand. A
   * registered client can do nothing at all until the household approves it,
   * so the gate is the consent page, not the registration endpoint.
   */
  get clientsStore(): OAuthRegisteredClientsStore {
    return {
      getClient: (clientId) => this.store.getClient(clientId),
      registerClient: (client) => this.store.putClient(client as OAuthClientInformationFull),
    };
  }

  /** Render the consent page. Issues nothing: only a click does that. */
  async authorize(
    client: OAuthClientInformationFull,
    params: AuthorizationParams,
    res: Response,
  ): Promise<void> {
    const identity = res.locals.identity as Identity | undefined;
    if (!identity) {
      // Belt and braces. The middleware in src/mcp/server.ts refuses first, so
      // getting here means the route was mounted without its gate.
      throw new ServerError(
        'the consent page was reached without an exe.dev identity. This is a routing ' +
          'mistake in the server, not something the client did wrong.',
      );
    }

    const consentId = this.desk.open(client, params, identity);
    res.setHeader('Cache-Control', 'no-store');
    res.setHeader(
      'Content-Security-Policy',
      "default-src 'none'; style-src 'unsafe-inline'; form-action 'self'",
    );
    res.setHeader('Referrer-Policy', 'no-referrer');
    res
      .status(200)
      .type('html')
      .send(
        consentPage({
          consentId,
          client,
          params,
          identity,
          folder: this.#folder,
          offerKroger: this.#kroger.configured,
          krogerConnected: this.#kroger.connected(),
        }),
      );
  }

  /**
   * Turn an approved consent into a code. Called from POST /consent, not by
   * the SDK — the SDK's flow stops at `authorize` and resumes at `/token`.
   */
  async issueCode(
    client: OAuthClientInformationFull,
    params: AuthorizationParams,
    identity: Identity,
  ): Promise<string> {
    if (!sameEmail(identity.email, this.owner)) {
      throw new ServerError(`${identity.email} is not the household`);
    }
    const code = newSecret();
    await this.store.putCode(code, {
      clientId: client.client_id,
      redirectUri: params.redirectUri,
      codeChallenge: params.codeChallenge,
      scopes: params.scopes ?? [],
      resource: params.resource?.href,
      subject: identity.email,
      expiresAt: nowSeconds() + CODE_TTL_SECONDS,
    });
    return code;
  }

  /**
   * The SDK calls this before exchangeAuthorizationCode, so it must READ the
   * code and leave it. Consuming it here would make every exchange fail.
   */
  async challengeForAuthorizationCode(
    client: OAuthClientInformationFull,
    authorizationCode: string,
  ): Promise<string> {
    const stored = this.store.getCode(authorizationCode);
    if (!stored || stored.clientId !== client.client_id) {
      throw new InvalidGrantError('the authorisation code is not valid, or has been used already');
    }
    if (stored.expiresAt <= nowSeconds()) {
      throw new InvalidGrantError('the authorisation code has expired. Start the flow again.');
    }
    return stored.codeChallenge;
  }

  async exchangeAuthorizationCode(
    client: OAuthClientInformationFull,
    authorizationCode: string,
    _codeVerifier?: string,
    redirectUri?: string,
    resource?: URL,
  ): Promise<OAuthTokens> {
    // Taken, not read: one code, one exchange. Two requests racing both call
    // take(), and only one of them gets the record.
    const stored = await this.store.takeCode(authorizationCode);
    if (!stored || stored.clientId !== client.client_id) {
      throw new InvalidGrantError('the authorisation code is not valid, or has been used already');
    }
    if (stored.expiresAt <= nowSeconds()) {
      throw new InvalidGrantError('the authorisation code has expired. Start the flow again.');
    }
    if (redirectUri !== undefined && redirectUri !== stored.redirectUri) {
      throw new InvalidGrantError(
        `redirect_uri "${redirectUri}" is not the one the code was issued for.`,
      );
    }
    if (
      resource !== undefined &&
      stored.resource !== undefined &&
      resource.href !== stored.resource
    ) {
      throw new InvalidGrantError(
        `this code was issued for ${stored.resource}, not ${resource.href}.`,
      );
    }

    return this.#issueTokens({
      clientId: client.client_id,
      scopes: stored.scopes,
      subject: stored.subject,
      resource: stored.resource,
    });
  }

  async exchangeRefreshToken(
    client: OAuthClientInformationFull,
    refreshToken: string,
    scopes?: string[],
    resource?: URL,
  ): Promise<OAuthTokens> {
    const stored = this.store.getRefreshToken(refreshToken);
    if (!stored || stored.clientId !== client.client_id) {
      throw new InvalidGrantError(
        'that refresh token is not valid. The household must approve again.',
      );
    }
    // A refresh may narrow the scopes it asks for. It may never widen them.
    const granted = scopes ?? stored.scopes;
    const widened = granted.filter((scope) => !stored.scopes.includes(scope));
    if (widened.length > 0) {
      throw new InvalidGrantError(
        `a refresh cannot ask for more than was approved. Not granted: ${widened.join(', ')}.`,
      );
    }

    // The old refresh token is retired with the exchange, so a stolen copy is
    // good for one use and the theft shows up as the real client being logged
    // out. RFC 9700 §4.14.2.
    await this.store.revokeRefreshToken(refreshToken);

    return this.#issueTokens({
      clientId: client.client_id,
      scopes: granted,
      subject: stored.subject,
      resource: resource?.href ?? stored.resource,
    });
  }

  async verifyAccessToken(token: string): Promise<AuthInfo> {
    const stored = this.store.getAccessToken(token);
    if (!stored) throw new InvalidTokenError('unknown access token');
    if (stored.expiresAt !== undefined && stored.expiresAt <= nowSeconds()) {
      throw new InvalidTokenError('the access token has expired');
    }
    // The store is the only thing that can put a subject on a token, and it
    // only ever gets one from a consent the owner gave. Checking it again here
    // costs nothing and means a change of owner takes effect on the next call
    // rather than at the next consent.
    if (!sameEmail(stored.subject, this.owner)) {
      throw new InvalidTokenError(
        `this token was issued to ${stored.subject}, and the meal plan belongs to ${this.owner}`,
      );
    }
    return {
      token,
      clientId: stored.clientId,
      scopes: stored.scopes,
      expiresAt: stored.expiresAt,
      resource: stored.resource === undefined ? undefined : new URL(stored.resource),
      extra: { email: stored.subject },
    };
  }

  async revokeToken(
    client: OAuthClientInformationFull,
    request: OAuthTokenRevocationRequest,
  ): Promise<void> {
    // RFC 7009: revoking a token that is already gone is a success, not an
    // error. Both bags are tried because the hint is only a hint.
    const access = this.store.getAccessToken(request.token);
    if (access && access.clientId === client.client_id) {
      await this.store.revokeAccessToken(request.token);
    }
    const refresh = this.store.getRefreshToken(request.token);
    if (refresh && refresh.clientId === client.client_id) {
      await this.store.revokeRefreshToken(request.token);
    }
  }

  async #issueTokens(grant: {
    clientId: string;
    scopes: string[];
    subject: string;
    resource?: string;
  }): Promise<OAuthTokens> {
    const accessToken = newSecret();
    const refreshToken = newSecret();
    await this.store.putAccessToken(accessToken, {
      ...grant,
      expiresAt: nowSeconds() + ACCESS_TOKEN_TTL_SECONDS,
    });
    await this.store.putRefreshToken(refreshToken, { ...grant });
    return {
      access_token: accessToken,
      token_type: 'Bearer',
      expires_in: ACCESS_TOKEN_TTL_SECONDS,
      scope: grant.scopes.join(' '),
      refresh_token: refreshToken,
    };
  }
}

/**
 * Real time, deliberately.
 *
 * The scenarios freeze a clock so git dates are deterministic, and that clock
 * must not reach token expiry: the SDK's requireBearerAuth compares expiresAt
 * against Date.now(), so a token minted on the frozen clock would be judged
 * against the real one. The scenario that needs an expired token expires it
 * through the store instead.
 */
function nowSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

File: lib/mealplan/auth/provider.ex
defmodule Mealplan.Auth.Provider do
  @moduledoc """
  This server is its own OAuth 2.1 authorisation server. Ported from
  `src/auth/provider.ts`.

  The SDK used to ship the HTTP half (`mcpAuthRouter`); on Elixir that half is
  `MealplanWeb.OAuthController`, and this module is the policy the SDK never
  shipped: who may consent, what a code means, how long a token lasts, and the
  PKCE and rotation rules.

  Tokens are opaque random strings, not JWTs — a JWT needs a signing key, one
  more secret in the process that already holds the household's credentials,
  and it turns revocation into a list of exceptions rather than a delete.
  `Mealplan.Auth.Store` keeps them SHA-256-hashed.

  The consent check is tenant membership ("owner of this tenant"), not one
  global email (plan 0005, Phase 3).
  """

  alias Mealplan.Accounts
  alias Mealplan.Auth.Store

  # RFC 6749 §4.1.2 says ten seconds; sixty is the generous end of that.
  @code_ttl_seconds 60

  @loopback_hosts ~w(localhost 127.0.0.1 [::1])

  # --- dynamic client registration -------------------------------------

  @doc """
  Register a client. `metadata` is the decoded JSON body. Registration is open
  by design — a registered client can do nothing until the household approves
  it, so the gate is the consent page, not this endpoint.
  """
  @spec register_client(map()) :: {:ok, map()} | {:error, String.t()}
  def register_client(metadata) when is_map(metadata) do
    redirect_uris = metadata["redirect_uris"]

    if is_list(redirect_uris) and redirect_uris != [] and Enum.all?(redirect_uris, &is_binary/1) do
      public? = metadata["token_endpoint_auth_method"] == "none"
      issued_at = Store.now_seconds()

      client =
        metadata
        |> Map.merge(%{
          "client_id" => uuid(),
          "client_id_issued_at" => issued_at
        })
        |> maybe_put_secret(public?, issued_at)

      {:ok, Store.put_client(client)}
    else
      {:error, "redirect_uris is required and must be a non-empty array of strings"}
    end
  end

  defp maybe_put_secret(client, true, _issued_at), do: client

  defp maybe_put_secret(client, false, _issued_at) do
    client
    |> Map.put("client_secret", :crypto.strong_rand_bytes(32) |> Base.encode16(case: :lower))
    # 0: registration secrets that expire would make the household re-register
    # an assistant every month for no gain.
    |> Map.put("client_secret_expires_at", 0)
  end

  @spec get_client(String.t()) :: map() | nil
  def get_client(client_id), do: Store.get_client(client_id)

  # --- the authorize request -----------------------------------------

  @doc """
  Validate an `/authorize` request. `query` has string keys.

  Returns `{:ok, %{client, redirect_uri, params}}` where `params` is the
  normalised authorization params, or:

  - `{:error, {:direct, message}}` — answer with 400, no redirect (bad
    client_id or redirect_uri; a redirect would be to an unverified URI);
  - `{:error, {:redirect, uri, error, description, state}}` — 302 to the
    redirect URI with the error parameters.
  """
  @spec validate_authorization(map()) ::
          {:ok, %{client: map(), redirect_uri: String.t(), params: map()}}
          | {:error, {:direct, String.t()}}
          | {:error, {:redirect, String.t(), String.t(), String.t(), String.t() | nil}}
  def validate_authorization(query) do
    client_id = query["client_id"]
    client = client_id && Store.get_client(client_id)

    cond do
      is_nil(client_id) or client_id == "" ->
        {:error, {:direct, "client_id is required"}}

      is_nil(client) ->
        {:error, {:direct, "Invalid client_id"}}

      true ->
        validate_redirect_and_params(client, query)
    end
  end

  defp validate_redirect_and_params(client, query) do
    registered = List.wrap(client["redirect_uris"])
    requested = query["redirect_uri"]

    redirect_uri =
      cond do
        is_binary(requested) -> requested
        length(registered) == 1 -> hd(registered)
        true -> nil
      end

    cond do
      is_nil(redirect_uri) ->
        {:error, {:direct, "redirect_uri is required when the client has more than one"}}

      not Enum.any?(registered, &redirect_uri_matches?(redirect_uri, &1)) ->
        {:error, {:direct, ~s(redirect_uri "#{redirect_uri}" is not registered for this client)}}

      true ->
        check_params(client, query, redirect_uri)
    end
  end

  defp check_params(client, query, redirect_uri) do
    state = query["state"]

    cond do
      query["response_type"] != "code" ->
        {:error,
         {:redirect, redirect_uri, "unsupported_response_type", "response_type must be \"code\"",
          state}}

      not is_binary(query["code_challenge"]) or query["code_challenge"] == "" ->
        {:error,
         {:redirect, redirect_uri, "invalid_request", "code_challenge is required (PKCE)", state}}

      query["code_challenge_method"] not in [nil, "S256"] ->
        {:error,
         {:redirect, redirect_uri, "invalid_request", "code_challenge_method must be S256", state}}

      true ->
        params = %{
          "redirect_uri" => redirect_uri,
          "code_challenge" => query["code_challenge"],
          "scopes" => split_scope(query["scope"]),
          "state" => state,
          "resource" => query["resource"]
        }

        {:ok, %{client: client, redirect_uri: redirect_uri, params: params}}
    end
  end

  # --- issuing a code, from the Approve click -----------------------

  @doc """
  Turn an approved consent into a code. Refuses unless `phone` is an owner of
  `tenant_slug` (ADR 0033 — identity is the telephone).
  """
  @spec issue_code(map(), map(), String.t(), String.t()) ::
          {:ok, String.t()} | {:error, String.t()}
  def issue_code(client, params, phone, tenant_slug) do
    if Accounts.owner?(tenant_slug, phone) do
      code = Store.new_secret()

      :ok =
        Store.put_code(code, %{
          client_id: client["client_id"],
          redirect_uri: params["redirect_uri"],
          code_challenge: params["code_challenge"],
          scopes: params["scopes"] || [],
          resource: params["resource"],
          subject: phone,
          expires_at: Store.now_seconds() + @code_ttl_seconds,
          tenant_id: tenant_id(tenant_slug)
        })

      {:ok, code}
    else
      {:error, "#{phone} does not own this meal plan"}
    end
  end

  # --- the token endpoint ------------------------------------------

  @doc """
  Exchange an authorization code. `params` has string keys: `code`,
  `code_verifier`, `redirect_uri`, `client_id`.
  """
  @spec exchange_authorization_code(map()) :: {:ok, map()} | {:error, {String.t(), String.t()}}
  def exchange_authorization_code(params) do
    client_id = params["client_id"]
    code = params["code"]

    # Taken, not read: one code, one exchange. Two racing requests both call
    # take_code/1 and only one gets the record.
    stored = code && Store.take_code(code)

    cond do
      is_nil(stored) or stored.client_id != client_id ->
        {:error,
         {"invalid_grant", "the authorisation code is not valid, or has been used already"}}

      stored.expires_at <= Store.now_seconds() ->
        {:error, {"invalid_grant", "the authorisation code has expired. Start the flow again."}}

      not pkce_ok?(params["code_verifier"], stored.code_challenge) ->
        {:error, {"invalid_grant", "the PKCE code_verifier does not match the code_challenge"}}

      is_binary(params["redirect_uri"]) and params["redirect_uri"] != stored.redirect_uri ->
        {:error,
         {"invalid_grant",
          ~s(redirect_uri "#{params["redirect_uri"]}" is not the one the code was issued for.)}}

      not resource_ok?(params["resource"], stored.resource) ->
        {:error,
         {"invalid_grant",
          "this code was issued for #{stored.resource}, not #{params["resource"]}."}}

      true ->
        {:ok,
         issue_tokens(%{
           client_id: client_id,
           scopes: stored.scopes || [],
           subject: stored.subject,
           resource: stored.resource,
           tenant_id: stored.tenant_id
         })}
    end
  end

  @doc "Exchange a refresh token, rotating it. `params`: `refresh_token`, `client_id`, `scope`, `resource`."
  @spec exchange_refresh_token(map()) :: {:ok, map()} | {:error, {String.t(), String.t()}}
  def exchange_refresh_token(params) do
    client_id = params["client_id"]
    refresh = params["refresh_token"]
    stored = refresh && Store.get_refresh_token(refresh)

    cond do
      is_nil(stored) or stored.client_id != client_id ->
        {:error,
         {"invalid_grant", "that refresh token is not valid. The household must approve again."}}

      true ->
        requested = if params["scope"], do: split_scope(params["scope"]), else: stored.scopes
        widened = requested -- (stored.scopes || [])

        if widened != [] do
          {:error,
           {"invalid_grant",
            "a refresh cannot ask for more than was approved. Not granted: #{Enum.join(widened, ", ")}."}}
        else
          # The old refresh token is retired with the exchange (RFC 9700
          # §4.14.2): a stolen copy is good for one use and the theft shows up
          # as the real client being logged out.
          :ok = Store.revoke_refresh_token(refresh)

          {:ok,
           issue_tokens(%{
             client_id: client_id,
             scopes: requested,
             subject: stored.subject,
             resource: params["resource"] || stored.resource,
             tenant_id: stored.tenant_id
           })}
        end
    end
  end

  # --- verifying a bearer token ----------------------------------

  @doc """
  Verify an access token. Returns `{:ok, auth_info}` or `{:error, message}`.
  `auth_info` carries `:client_id`, `:scopes`, `:expires_at`, `:subject`,
  `:resource`, `:tenant_id` and `:tenant` (the slug).

  The token is trusted for the tenant it was ISSUED for — `stored.tenant_id`,
  set when the code was minted — and then the telephone `subject` is re-checked
  against the store for still owning that tenant (ADR 0033). A membership
  removed by `mix mealplan.invite --revoke` fails here on the next call, so
  revoking an invitation logs that household's clients out within one request.
  """
  @spec verify_access_token(String.t()) :: {:ok, map()} | {:error, String.t()}
  def verify_access_token(token) do
    stored = Store.get_access_token(token)

    cond do
      is_nil(stored) ->
        {:error, "unknown access token"}

      not is_nil(stored.expires_at) and stored.expires_at <= Store.now_seconds() ->
        {:error, "the access token has expired"}

      not Accounts.owner_of_tenant_id?(stored.tenant_id, stored.subject) ->
        {:error, "this token was issued to #{stored.subject}, who no longer owns this meal plan"}

      true ->
        {:ok,
         %{
           token: token,
           client_id: stored.client_id,
           scopes: stored.scopes || [],
           expires_at: stored.expires_at,
           resource: stored.resource,
           subject: stored.subject,
           tenant_id: stored.tenant_id,
           tenant: tenant_slug(stored.tenant_id)
         }}
    end
  end

  @doc "RFC 7009: revoking a token that is already gone is a success. The hint is only a hint."
  @spec revoke_token(String.t(), String.t()) :: :ok
  def revoke_token(client_id, token) do
    case Store.get_access_token(token) do
      %{client_id: ^client_id} -> Store.revoke_access_token(token)
      _ -> :ok
    end

    case Store.get_refresh_token(token) do
      %{client_id: ^client_id} -> Store.revoke_refresh_token(token)
      _ -> :ok
    end

    :ok
  end

  # --- helpers -------------------------------------------------

  @doc """
  RFC 8252 §7.3: an authorisation server MUST allow any port for a loopback
  redirect URI. For everything else, exact match.
  """
  def redirect_uri_matches?(requested, registered) do
    requested == registered or loopback_port_relaxed?(requested, registered)
  end

  defp loopback_port_relaxed?(requested, registered) do
    with %URI{} = req <- URI.parse(requested),
         %URI{} = reg <- URI.parse(registered),
         true <- req.host in @loopback_hosts and reg.host in @loopback_hosts do
      req.scheme == reg.scheme and req.host == reg.host and (req.path || "/") == (reg.path || "/") and
        req.query == reg.query
    else
      _ -> false
    end
  end

  defp pkce_ok?(verifier, challenge) when is_binary(verifier) and is_binary(challenge) do
    computed = :crypto.hash(:sha256, verifier) |> Base.url_encode64(padding: false)
    Plug.Crypto.secure_compare(computed, challenge)
  end

  defp pkce_ok?(_, _), do: false

  defp resource_ok?(nil, _stored), do: true
  defp resource_ok?(_requested, nil), do: true
  defp resource_ok?(requested, stored), do: requested == stored

  defp split_scope(nil), do: []
  defp split_scope(""), do: []
  defp split_scope(scope) when is_binary(scope), do: String.split(scope, ~r/\s+/, trim: true)

  defp issue_tokens(grant) do
    access = Store.new_secret()
    refresh = Store.new_secret()
    ttl = Store.access_token_ttl_seconds()

    :ok =
      Store.put_access_token(access, %{
        client_id: grant.client_id,
        scopes: grant.scopes,
        subject: grant.subject,
        resource: grant.resource,
        tenant_id: grant.tenant_id,
        expires_at: Store.now_seconds() + ttl
      })

    :ok =
      Store.put_refresh_token(refresh, %{
        client_id: grant.client_id,
        scopes: grant.scopes,
        subject: grant.subject,
        resource: grant.resource,
        tenant_id: grant.tenant_id
      })

    %{
      "access_token" => access,
      "token_type" => "Bearer",
      "expires_in" => ttl,
      "scope" => Enum.join(grant.scopes, " "),
      "refresh_token" => refresh
    }
  end

  defp tenant_id(tenant_slug) do
    case Accounts.get_tenant_by_slug(tenant_slug) do
      %{id: id} -> id
      _ -> nil
    end
  end

  defp tenant_slug(nil), do: nil

  defp tenant_slug(tenant_id) do
    case Mealplan.Repo.get(Mealplan.Accounts.Tenant, tenant_id) do
      %{slug: slug} -> slug
      _ -> nil
    end
  end

  defp uuid do
    <<a::32, b::16, c::16, d::16, e::48>> = :crypto.strong_rand_bytes(16)
    c = Bitwise.bor(Bitwise.band(c, 0x0FFF), 0x4000)
    d = Bitwise.bor(Bitwise.band(d, 0x3FFF), 0x8000)

    :io_lib.format("~8.16.0b-~4.16.0b-~4.16.0b-~4.16.0b-~12.16.0b", [a, b, c, d, e])
    |> IO.iodata_to_binary()
  end
end
